crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
   }
 
-  if (*siglen < DILITHIUM3_SIGNATURE_BYTES) {
+  if (*siglen < nistdsa->signature_len) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
     return 0;
   }
 
-  if (ml_dsa_65_sign(sig, siglen, tbs, tbslen, NULL, 0, key->priv) != 0) {
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.nistdsa_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+      }
+
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
+  if (!key->secret_key) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (nistdsa->method->sign(sig, siglen, tbs, tbslen, NULL,
+                            0, key->secret_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, ERR_R_INTERNAL_ERROR);
     return 0;
   }
   return 1;
 }
 
-static int pkey_dilithium3_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
+static int pkey_nistdsa_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
                                           size_t siglen, const uint8_t *tbs,
                                           size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
-    return 0;
+  NISTDSA_PKEY_CTX *dctx = ctx->data;
+  const NISTDSA *nistdsa = dctx->nistdsa;
+
+  if (nistdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+
+    nistdsa = NISTDSA_KEY_get0_sig(ctx->pkey->pkey.nistdsa_key);
   }
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+    ctx->pkey->pkey.nistdsa_key == NULL ||
+    ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+    }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
 
-  if (siglen != DILITHIUM3_SIGNATURE_BYTES ||
-      ml_dsa_65_verify(tbs, tbslen, sig, siglen, NULL, 0, key->pub) != 0) {
+  if (siglen != nistdsa->signature_len ||
+      nistdsa->method->verify(tbs, tbslen, sig, siglen,
+                              NULL, 0, key->public_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
+    return 0;
+      }
+
+  return 1;
+}
+
+// This function sets nistdsa parameters defined by |nid| in |pkey|.
+// If |pkey| already has a public key set, this public key is preserved.
+int EVP_PKEY_nistdsa_set_params(EVP_PKEY *pkey, int nid) {
+  const NISTDSA *nistdsa = SIG_find_dsa_by_nid(nid);
+
+  if (nistdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+
+  // if the public key has already been set either by EVP_parse_public_key or
+  // some other method that returns a PKEY without setting params, then
+  // we preserve that PKEY and just populate the params
+  if (pkey->pkey.nistdsa_key != NULL) {
+    pkey->pkey.nistdsa_key->nistdsa = nistdsa;
```

Comment:
looks like you also assume that the asn1 method is set in that case?

Reply:
Yes, in fact, the entire point of this if statement is to circumvent an issue that calling `evp_pkey_set_method(pkey, &pqdsa_asn1_meth)` creates. Let me describe:

Often we find ourselves in a situation in which we have a PKEY with some key material set, but the `pkey->pqdsa_key->pqdsa` struct is NULL (i.e., no asn1, pkey, or other params set). The function `EVP_PKEY_pqdsa_set_params` is designed to set the `pqdsa` struct for the correct NID, however, when we call `evp_pkey_set_method(pkey, &pqdsa_asn1_meth)` on line 159, one of the first things `evp_pkey_set_method` does is free the `pkey` (see 
https://github.com/jakemas/aws-lc/blob/8d6ff48bbe5888bab04acccadf12eca319fb45ba/crypto/fipsmodule/evp/evp.c#L261).

Thus, `evp_pkey_set_method` correctly set the methods/params, but wiped the incoming `PKEY` that had some key material set. So, to avoid calling `evp_pkey_set_method` we interrupt `EVP_PKEY_pqdsa_set_params` for the case in which key material is set, but methods/params are not.

Reply:
To address https://github.com/aws/aws-lc/pull/1963#discussion_r1831196097 and https://github.com/aws/aws-lc/pull/1963#discussion_r1831588720 I've changed the implementation of `EVP_PKEY_pqdsa_set_params`. See more at https://github.com/aws/aws-lc/pull/1963#discussion_r1833473845 and https://github.com/aws/aws-lc/pull/1963/commits/c0d4e651b5accb0d1a05fc260635ce1d8968b3e8